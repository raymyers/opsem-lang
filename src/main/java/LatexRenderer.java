import ast.*;
import org.stringtemplate.v4.ST;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class LatexRenderer {

    private final Program program;
    private final Map<String, LatexRendering> renderingMap;

    public LatexRenderer(Program program) {
        this.program = program;
        this.renderingMap = new HashMap<>();
        this.program.latexRenderings().forEach(r ->
                this.renderingMap.put(r.label(), r));

    }
    public static String programToLatex(Program program) {
        return new LatexRenderer(program).toLatex();
    }

    public String toLatex() {
        return program.rules().stream().map(this::semRuleToLatex)
                .collect(Collectors.joining("\n\n"));
    }

    public String condToLatex(Cond cond) {
        String inner = cond.exprs().stream().map(this::exprsToLatex).collect(Collectors.joining(" ,\\, "));
        if (cond.exprs().size() == 1) {
            return inner;
        }
        return "(" + inner + ")";
    }

    public String simpleCondLineToLatex(CondLine.SimpleCondLine condLine) {
        if (condLine.post().isPresent()) {
            return condToLatex(condLine.pre()) + " => " + condToLatex(condLine.post().get());
        }
        return condToLatex(condLine.pre());
    }

    public String varToLatex(Expr.Var var) {
        String name = var.name();
        if (this.renderingMap.containsKey(name)) {
            LatexRendering latexRendering = this.renderingMap.get(name);
            if (latexRendering.params().isEmpty()) {
                return latexRendering.rendering();
            }
        }
        return trailingNumberToSubscript(name);
    }

    private static String trailingNumberToSubscript(String name) {
        Pattern endWithDigits = Pattern.compile(".*(\\D)(\\d+)$");

        Matcher matcher = endWithDigits.matcher(name);
        if (matcher.matches()) {
            String base = matcher.group(1);
            String sub = matcher.group(2);
            return base + "_" + sub;
        }
        return name;
    }

    public String exprsWrapToLatex(Expr.ExprsWrap exprsWrap) {
        return "(" + exprsToLatex(exprsWrap.exprs()) + ")";
    }

    public String exprsToLatex(Exprs exprs) {
        if (exprs.exprs().size() > 1) {
            Expr first = exprs.exprs().get(0);
            int argCount = exprs.exprs().size() - 1;
            if (first instanceof Expr.Var label) {
                LatexRendering latexRendering = this.renderingMap.get(label.name());
                if (latexRendering != null && latexRendering.params().size() == argCount) {
                    ST template = new ST(latexRendering.rendering());
                    for (int i = 0; i < argCount; i++) {
                        Expr arg = exprs.exprs().get(i + 1);
                        String param = latexRendering.params().get(i);
                        template.add(param, exprToLatex(arg));
                    }
                    return template.render();
                } else {
                    // WARN?
                    System.err.println("WARN: Arg count mismatch for label " + label.name());
                }
            }
        }
        return exprs.exprs().stream().map(this::exprToLatex).collect(Collectors.joining("\\,"));
    }

    private String exprToLatex(Expr e) {
        if (e instanceof Expr.ExprsWrap ew) return exprsWrapToLatex(ew);
        if (e instanceof Expr.Var v) return varToLatex(v);
        return "";
    }

    public String semRuleToLatex(SemRule semRule) {
        return slashBracketWrap(condLayersToLatex(semRule.name(), semRule.layers()));
    }

    private String condLayersToLatex(String name, List<CondLayer> layers) {
        String templateStr = "\\inference {<top>}{<bottom>}[<name>]";
        ST template = new ST(templateStr);
        boolean topFull = false;
        boolean bottomFull = false;
        String sep = " & ";

        for (var layer : layers) {
            String renderedLayer = condLayerToLatex(layer, sep);
            if (!topFull) {
                template.add("top", renderedLayer);
                topFull = true;
            } else if (!bottomFull) {
                template.add("bottom", renderedLayer);
                bottomFull = true;
            } else {
                ST innerTemplate = template;
                template = new ST(templateStr);
                template.add("top", innerTemplate.render());
                template.add("bottom", renderedLayer);
            }
        }
        template.add("name", name);

        return template.render();
    }

    private static String slashBracketWrap(String inference) {
        return "\\[\n" + inference + "\n\\]";
    }

    private String condLayerToLatex(CondLayer layer, String sep) {
        return layer.condLines().stream().map(this::condLineToLatex).collect(Collectors.joining(sep));
    }

    private String condLineToLatex(CondLine condLine) {
        if (condLine instanceof CondLine.SimpleCondLine simpleCondLine) {
            return simpleCondLineToLatex(simpleCondLine);
        }
        if (condLine instanceof CondLine.CondBlock condBlock) {
            return condLayersToLatex("", condBlock.layers());
        }
        return "";
    }


}
