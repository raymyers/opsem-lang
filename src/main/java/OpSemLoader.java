import ast.Program;
import org.antlr.v4.runtime.*;

import java.util.ArrayList;
import java.util.List;

public class OpSemLoader {

    private final List<String> syntaxErrors;
    private final OpSemParser parser;
    private final OpSemParser.StartContext startContext;

    private class ErrorListener extends ConsoleErrorListener {

        private final List<String> syntaxErrors = new ArrayList<>();

        @Override
        public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int line, int charPositionInLine, String msg, RecognitionException e) {
            syntaxErrors.add(String.format("Syntax Error on %s:%s - %s", line, charPositionInLine, msg));
        }

        public List<String> getSyntaxErrors(){
            return syntaxErrors;
        }
    }

    OpSemLoader(String input){
        OpSemLexer lexer = new OpSemLexer(CharStreams.fromString(input));
        this.parser = new OpSemParser(new CommonTokenStream(lexer));

        ErrorListener errorListener = new ErrorListener();

        //Remove and Replace Default Console Error Listeners
        lexer.removeErrorListeners();
        parser.removeErrorListeners();
        parser.addErrorListener(errorListener);
        lexer.addErrorListener(errorListener);

        syntaxErrors = errorListener.getSyntaxErrors();
        startContext = parser.start();
    }

    private String getStringTree() {
        return startContext.toStringTree(parser);
    }

    public Program toProgram() {
        return new OpSemTransformer().transformStart(startContext);
    }

    public boolean valid(){
        return syntaxErrors.size() == 0;
    }

    public String getSyntaxErrors(){
        return String.join("\n",syntaxErrors) + "\n";
    }
}
