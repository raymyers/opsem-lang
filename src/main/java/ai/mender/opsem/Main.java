package ai.mender.opsem;

import picocli.CommandLine;
import picocli.CommandLine.Command;
import picocli.CommandLine.Parameters;

import java.io.File;
import java.util.concurrent.Callable;

@Command(name = "opsem", mixinStandardHelpOptions = true, version = "opsem 0.0.0",
         description = "Generate LaTex from opsem")
class Main implements Callable<Integer> {

    @Parameters(index = "0", description = "The file to process.")
    private File file;

//    @Option(names = {"-a", "--algorithm"}, description = "MD5, SHA-1, SHA-256, ...")
//    private String algorithm = "SHA-256";

    @Override
    public Integer call() throws Exception {
        OpSemLoader subject = new OpSemLoader(file);
        if (!subject.valid()) {
            System.err.println(subject.getSyntaxErrors());
            return 1;
        }

        System.out.println(LatexRenderer.programToLatex(subject.toProgram()));
        return 0;
    }

    // this example implements Callable, so parsing, error handling and handling user
    // requests for usage help or version help can be done with one line of code.
    public static void main(String... args) {
        int exitCode = new CommandLine(new Main()).execute(args);
        System.exit(exitCode);
    }
}