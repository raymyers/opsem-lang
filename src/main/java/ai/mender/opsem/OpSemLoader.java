package ai.mender.opsem;

import ai.mender.opsem.generated.OpSemLexer;
import ai.mender.opsem.generated.OpSemParser;
import ast.Program;
import org.antlr.v4.runtime.*;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class OpSemLoader {

    private final List<String> syntaxErrors;
    private final OpSemParser parser;
    private final OpSemParser.StartContext startContext;

    public OpSemLoader(File file) throws IOException {
        this(new OpSemLexer(CharStreams.fromPath(file.toPath())));
    }
    public OpSemLoader(String input){
        this(new OpSemLexer(CharStreams.fromString(input)));
    }


    private OpSemLoader(OpSemLexer lexer) {
        this.parser = new OpSemParser(new CommonTokenStream(lexer));

        ErrorListener errorListener = new ErrorListener();

        //Remove and Replace Default Console Error Listeners
        lexer.removeErrorListeners();
        parser.removeErrorListeners();
        parser.addErrorListener(errorListener);
        lexer.addErrorListener(errorListener);

        syntaxErrors = errorListener.getSyntaxErrors();
        startContext = parser.start();
    }

    private class ErrorListener extends ConsoleErrorListener {

        private final List<String> syntaxErrors = new ArrayList<>();

        @Override
        public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int line, int charPositionInLine, String msg, RecognitionException e) {
            syntaxErrors.add(String.format("Syntax Error on %s:%s - %s", line, charPositionInLine, msg));
        }

        public List<String> getSyntaxErrors(){
            return syntaxErrors;
        }
    }


    private String getStringTree() {
        return startContext.toStringTree(parser);
    }

    public Program toProgram() {
        return new OpSemTransformer().transformStart(startContext);
    }

    public boolean valid(){
        return syntaxErrors.size() == 0;
    }

    public String getSyntaxErrors(){
        return String.join("\n",syntaxErrors) + "\n";
    }
}
