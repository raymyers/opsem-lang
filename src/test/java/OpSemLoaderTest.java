import ai.mender.opsem.LatexRenderer;
import ai.mender.opsem.OpSemLoader;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class OpSemLoaderTest {


    public static final String FULL_PROGRAM_1 = """
        rule Skip {
            ~
            Skip, s => s
        }
                
        rule Assign {
            ~
            Assign x a, s => Put s (Eval a s)
        }
                
        rule Seq {
            c1, s1 => s2
            c2, s2 => s3
            ~
            Seq c1 c2, s1 => s3
        }
                
        rule WhileTrue {
            bval b s1
            c, s1 => s2
            While b c, s2 => s3
            ~
            While b c, s1 => s3
        }
    """;

    public static final String FULL_PROGRAM_2 = """
        latex {
            context = "\\Gamma"
            WellTyped c a = "<c> \\vdash <a>"
            HasType c a b = "<c> \\vdash <a> : <b>"
            TypeOf c a = "<c> <a>"
            While b c = "\\text{WHILE } <b> \\text{ DO } <c>"
            Assign a b = "<a> ::= <b>"
            Seq a b = "<a> \\text{ ;; } <b>"
            If a b c = "\\text{IF }a  \\text{ THEN } b \\text{ ELSE } b)"
        }
        
        rule SkipT {
            ~
            WellTyped context SKIP
        }
                
        rule AssignT {
            HasType context a (TypeOf context x)
            ~
            WellTyped context (Assign x a)
        }
                
        rule SeqT {
            WellTyped context c1
            WellTyped context c2
            ~
            WellTyped context (Seq c1 c2)
        }
        rule IfT {
            WellTyped context b
            WellTyped context c1
            WellTyped context c2
            ~
            WellTyped context (If b c1 c2)
        }
        rule WhileT {
            WellTyped context b
            WellTyped context c
            ~
            WellTyped context (While b c)
        }
    """;

    public static final String TREE = """
        latex {
            WellTyped a = "\\Gamma \\vdash <a>"
            HasType a b = "\\Gamma \\vdash <a> : <b>"
            TypeOf a = "\\Gamma\\,<a>"
            While b c = "\\text{WHILE } <b> \\text{ DO } <c>"
            Assign a b = "<a> ::= <b>"
            Seq a b = "<a> \\text{ ;; } <b>"
            If a b c = "\\text{IF }a  \\text{ THEN } b \\text{ ELSE } b)"
            Eq a b = "<a> = <b>"
        }

        rule Derivation912 {
            {
                Eq (TypeOf y) Ity
                ~
                HasType (V x) Ity
                ~
                WellTyped (Assign x (V y))
            }
            {
                {
                    Eq (TypeOf x) Ity
                    ~
                    HasType (V x) Ity
                }
                {
                    Eq (TypeOf y) Ity
                    ~
                    HasType (V y) Ity
                }
                ~
                HasType (Plus (V x) (V y)) Ity
                ~
                WellTyped (Assign y (Plus (V x) (V y)))
            }
            ~
            WellTyped (Seq (Assign x (V y)) (Plus (V x) (V y)))
        }
    """;

    @Test
    void generatAst()
    {
        OpSemLoader subject = new OpSemLoader(FULL_PROGRAM_1);
        assertTrue(subject.valid());

        ApprovalHelpers.verifyAsJson(subject.toProgram());
    }

    @Test
    void toLatex()
    {
        String input = """                    
            rule WhileTrue {
                bval b s1
                c, s1 => s2
                While b c, s2 => s3
                ~
                While b c, s1 => s3
            }
        """;
        OpSemLoader subject = new OpSemLoader(input);
        assertTrue(subject.valid(), subject.getSyntaxErrors());

        String latex = new LatexRenderer(subject.toProgram()).semRuleToLatex(subject.toProgram().rules().get(0));
        Approvals.verify(latex);
    }


    @Test
    void toLatexFullProgram()
    {
        OpSemLoader subject = new OpSemLoader(FULL_PROGRAM_1);
        assertTrue(subject.valid(), subject.getSyntaxErrors());

        String latex = LatexRenderer.programToLatex(subject.toProgram());
        Approvals.verify(latex);
    }


    @Test
    void astFullProgramWithRenderings()
    {
        OpSemLoader subject = new OpSemLoader(FULL_PROGRAM_2);
        assertTrue(subject.valid(), subject.getSyntaxErrors());

        ApprovalHelpers.verifyAsJson(subject.toProgram());
    }
    @Test
    void toLatexFullProgramWithRenderings()
    {
        OpSemLoader subject = new OpSemLoader(FULL_PROGRAM_2);
        assertTrue(subject.valid(), subject.getSyntaxErrors());

        String latex = LatexRenderer.programToLatex(subject.toProgram());
        Approvals.verify(latex);
    }
    @Test
    void toLatexDerivationTree()
    {
        OpSemLoader subject = new OpSemLoader(TREE);
        assertTrue(subject.valid(), subject.getSyntaxErrors());

        String latex = LatexRenderer.programToLatex(subject.toProgram());
        Approvals.verify(latex);
    }
}
