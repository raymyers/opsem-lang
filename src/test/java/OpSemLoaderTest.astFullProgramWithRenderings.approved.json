{
  "@type" : "Program",
  "rules" : [ {
    "@type" : "SemRule",
    "name" : "SkipT",
    "topLines" : [ ],
    "bottomLines" : [ {
      "@type" : "CondLine",
      "pre" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "WellTyped"
          }, {
            "@type" : "Var",
            "name" : "context"
          }, {
            "@type" : "Var",
            "name" : "SKIP"
          } ]
        } ]
      },
      "post" : null
    } ]
  }, {
    "@type" : "SemRule",
    "name" : "AssignT",
    "topLines" : [ {
      "@type" : "CondLine",
      "pre" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "HasType"
          }, {
            "@type" : "Var",
            "name" : "context"
          }, {
            "@type" : "Var",
            "name" : "a"
          }, {
            "@type" : "ExprsWrap",
            "exprs" : {
              "exprs" : [ {
                "@type" : "Var",
                "name" : "TypeOf"
              }, {
                "@type" : "Var",
                "name" : "context"
              }, {
                "@type" : "Var",
                "name" : "x"
              } ]
            }
          } ]
        } ]
      },
      "post" : null
    } ],
    "bottomLines" : [ {
      "@type" : "CondLine",
      "pre" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "WellTyped"
          }, {
            "@type" : "Var",
            "name" : "context"
          }, {
            "@type" : "ExprsWrap",
            "exprs" : {
              "exprs" : [ {
                "@type" : "Var",
                "name" : "Assign"
              }, {
                "@type" : "Var",
                "name" : "x"
              }, {
                "@type" : "Var",
                "name" : "a"
              } ]
            }
          } ]
        } ]
      },
      "post" : null
    } ]
  }, {
    "@type" : "SemRule",
    "name" : "SeqT",
    "topLines" : [ {
      "@type" : "CondLine",
      "pre" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "WellTyped"
          }, {
            "@type" : "Var",
            "name" : "context"
          }, {
            "@type" : "Var",
            "name" : "c1"
          } ]
        } ]
      },
      "post" : null
    }, {
      "@type" : "CondLine",
      "pre" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "WellTyped"
          }, {
            "@type" : "Var",
            "name" : "context"
          }, {
            "@type" : "Var",
            "name" : "c2"
          } ]
        } ]
      },
      "post" : null
    } ],
    "bottomLines" : [ {
      "@type" : "CondLine",
      "pre" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "WellTyped"
          }, {
            "@type" : "Var",
            "name" : "context"
          }, {
            "@type" : "ExprsWrap",
            "exprs" : {
              "exprs" : [ {
                "@type" : "Var",
                "name" : "Seq"
              }, {
                "@type" : "Var",
                "name" : "c1"
              }, {
                "@type" : "Var",
                "name" : "c2"
              } ]
            }
          } ]
        } ]
      },
      "post" : null
    } ]
  }, {
    "@type" : "SemRule",
    "name" : "IfT",
    "topLines" : [ {
      "@type" : "CondLine",
      "pre" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "WellTyped"
          }, {
            "@type" : "Var",
            "name" : "context"
          }, {
            "@type" : "Var",
            "name" : "b"
          } ]
        } ]
      },
      "post" : null
    }, {
      "@type" : "CondLine",
      "pre" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "WellTyped"
          }, {
            "@type" : "Var",
            "name" : "context"
          }, {
            "@type" : "Var",
            "name" : "c1"
          } ]
        } ]
      },
      "post" : null
    }, {
      "@type" : "CondLine",
      "pre" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "WellTyped"
          }, {
            "@type" : "Var",
            "name" : "context"
          }, {
            "@type" : "Var",
            "name" : "c2"
          } ]
        } ]
      },
      "post" : null
    } ],
    "bottomLines" : [ {
      "@type" : "CondLine",
      "pre" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "WellTyped"
          }, {
            "@type" : "Var",
            "name" : "context"
          }, {
            "@type" : "ExprsWrap",
            "exprs" : {
              "exprs" : [ {
                "@type" : "Var",
                "name" : "If"
              }, {
                "@type" : "Var",
                "name" : "b"
              }, {
                "@type" : "Var",
                "name" : "c1"
              }, {
                "@type" : "Var",
                "name" : "c2"
              } ]
            }
          } ]
        } ]
      },
      "post" : null
    } ]
  }, {
    "@type" : "SemRule",
    "name" : "WhileT",
    "topLines" : [ {
      "@type" : "CondLine",
      "pre" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "WellTyped"
          }, {
            "@type" : "Var",
            "name" : "context"
          }, {
            "@type" : "Var",
            "name" : "b"
          } ]
        } ]
      },
      "post" : null
    }, {
      "@type" : "CondLine",
      "pre" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "WellTyped"
          }, {
            "@type" : "Var",
            "name" : "context"
          }, {
            "@type" : "Var",
            "name" : "c"
          } ]
        } ]
      },
      "post" : null
    } ],
    "bottomLines" : [ {
      "@type" : "CondLine",
      "pre" : {
        "@type" : "Cond",
        "exprs" : [ {
          "exprs" : [ {
            "@type" : "Var",
            "name" : "WellTyped"
          }, {
            "@type" : "Var",
            "name" : "context"
          }, {
            "@type" : "ExprsWrap",
            "exprs" : {
              "exprs" : [ {
                "@type" : "Var",
                "name" : "While"
              }, {
                "@type" : "Var",
                "name" : "b"
              }, {
                "@type" : "Var",
                "name" : "c"
              } ]
            }
          } ]
        } ]
      },
      "post" : null
    } ]
  } ],
  "latexRenderings" : [ {
    "label" : "context",
    "params" : [ ],
    "rendering" : "\\Gamma"
  }, {
    "label" : "WellTyped",
    "params" : [ "c", "a" ],
    "rendering" : "<c> \\vdash <a>"
  }, {
    "label" : "HasType",
    "params" : [ "c", "a", "b" ],
    "rendering" : "<c> \\vdash <a> : <b>"
  }, {
    "label" : "TypeOf",
    "params" : [ "c", "a" ],
    "rendering" : "<c> <a>"
  }, {
    "label" : "While",
    "params" : [ "b", "c" ],
    "rendering" : "\\text{WHILE } <b> \\text{ DO } <c>"
  }, {
    "label" : "Assign",
    "params" : [ "a", "b" ],
    "rendering" : "<a> ::= <b>"
  }, {
    "label" : "Seq",
    "params" : [ "a", "b" ],
    "rendering" : "<a> \\text{ ;; } <b>"
  }, {
    "label" : "If",
    "params" : [ "a", "b", "c" ],
    "rendering" : "\\text{IF }a  \\text{ THEN } b \\text{ ELSE } b)"
  } ]
}